File: backend/app/api/v1/endpoints/simulator.py
# =============================================================================
# ADs Growth System - ML Simulator Endpoints
# =============================================================================
"""
What-If Simulator and ML prediction endpoints.
Implements Module A: Intelligence Engine.
"""

from datetime import datetime, timezone
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Request, status
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.auth.deps import get_current_user
from app.core.config import settings
from app.core.logging import get_logger
from app.db.session import get_async_session
from app.models import Campaign
from app.schemas import (
    APIResponse,
    ConversionPredictionRequest,
    ConversionPredictionResponse,
    ROASForecastRequest,
    ROASForecastResponse,
    SimulationRequest,
    SimulationResponse,
)

logger = get_logger(__name__)

# STRAT-SC-001: router-level auth — the registration-time FeatureGate is an
# env kill-switch, not authentication (same fail-open class C3/C6 closed on
# 15 sibling routers).
router = APIRouter(dependencies=[Depends(get_current_user)])


@router.post("", response_model=APIResponse[SimulationResponse])
async def simulate_budget_change(
    simulation: SimulationRequest,
    db: AsyncSession = Depends(get_async_session),
):
    """
    What-If Simulator: Predict outcomes based on budget changes.

    Uses the hybrid ML strategy:
    - ML_PROVIDER=local: Uses scikit-learn models
    - ML_PROVIDER=vertex: Uses Google Vertex AI
    """
    # Get campaign data if specified
    campaign_data = None
    if simulation.campaign_id:
        result = await db.execute(
            select(Campaign).where(
                Campaign.id == simulation.campaign_id,
            )
        )
        campaign = result.scalar_one_or_none()

        if not campaign:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Campaign not found",
            )

        campaign_data = {
            "current_spend": campaign.total_spend_cents / 100,
            "impressions": campaign.impressions,
            "clicks": campaign.clicks,
            "conversions": campaign.conversions,
            "revenue": campaign.revenue_cents / 100,
            "ctr": campaign.ctr,
            "roas": campaign.roas,
        }
    else:
        # Use aggregated portfolio data across all campaigns
        result = await db.execute(
            select(Campaign).where(
                Campaign.is_deleted == False,
            )
        )
        campaigns = result.scalars().all()

        total_spend = sum(c.total_spend_cents for c in campaigns) / 100
        total_revenue = sum(c.revenue_cents for c in campaigns) / 100
        total_impressions = sum(c.impressions for c in campaigns)
        total_clicks = sum(c.clicks for c in campaigns)
        total_conversions = sum(c.conversions for c in campaigns)

        campaign_data = {
            "current_spend": total_spend,
            "impressions": total_impressions,
            "clicks": total_clicks,
            "conversions": total_conversions,
            "revenue": total_revenue,
            "ctr": (
                (total_clicks / total_impressions * 100) if total_impressions > 0 else 0
            ),
            "roas": (total_revenue / total_spend) if total_spend > 0 else 0,
        }

    # Run prediction using the ML service
    from app.ml.simulator import WhatIfSimulator

    simulator = WhatIfSimulator()

    prediction = await simulator.predict_budget_impact(
        current_metrics=campaign_data,
        budget_change_percent=simulation.budget_change_percent,
        days_ahead=simulation.days_ahead,
        include_confidence=simulation.include_confidence_interval,
    )

    return APIResponse(
        success=True,
        data=SimulationResponse(
            campaign_id=simulation.campaign_id,
            current_metrics=campaign_data,
            predicted_metrics=prediction["predicted_metrics"],
            budget_change_percent=simulation.budget_change_percent,
            confidence_interval=prediction.get("confidence_interval"),
            feature_importances=prediction.get("feature_importances"),
            model_version=prediction["model_version"],
        ),
    )


@router.post("/forecast/roas", response_model=APIResponse[ROASForecastResponse])
async def forecast_roas(
    forecast_request: ROASForecastRequest,
    db: AsyncSession = Depends(get_async_session),
):
    """
    ROAS Forecasting: Predict future ROAS trends.

    Uses historical data to forecast ROAS for the specified period.
    """
    # Get campaigns to forecast
    query = select(Campaign).where(
        Campaign.is_deleted == False,
    )

    if forecast_request.campaign_ids:
        query = query.where(Campaign.id.in_(forecast_request.campaign_ids))

    result = await db.execute(query)
    campaigns = result.scalars().all()

    if not campaigns:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="No campaigns found",
        )

    # Get historical metrics for forecasting
    from app.ml.forecaster import ROASForecaster

    forecaster = ROASForecaster()

    forecasts = await forecaster.forecast(
        campaigns=campaigns,
        days_ahead=forecast_request.days_ahead,
        granularity=forecast_request.granularity,
        db=db,
    )

    return APIResponse(
        success=True,
        data=ROASForecastResponse(
            forecasts=forecasts["predictions"],
            model_version=forecasts["model_version"],
            generated_at=datetime.now(timezone.utc),
        ),
    )


@router.post(
    "/predict/conversions", response_model=APIResponse[ConversionPredictionResponse]
)
async def predict_conversions(
    prediction_request: ConversionPredictionRequest,
    db: AsyncSession = Depends(get_async_session),
):
    """
    Conversion Prediction: Predict expected conversions based on features.

    Can be used to estimate conversions for different targeting scenarios.
    """
    # Verify campaign exists
    result = await db.execute(
        select(Campaign).where(
            Campaign.id == prediction_request.campaign_id,
        )
    )
    campaign = result.scalar_one_or_none()

    if not campaign:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Campaign not found",
        )

    from app.ml.conversion_predictor import ConversionPredictor

    predictor = ConversionPredictor()

    # Merge campaign features with custom features
    features = {
        "impressions": campaign.impressions,
        "clicks": campaign.clicks,
        "spend": campaign.total_spend_cents / 100,
        "ctr": campaign.ctr or 0,
        "platform": campaign.platform.value,
        **prediction_request.features,
    }

    prediction = await predictor.predict(features)

    return APIResponse(
        success=True,
        data=ConversionPredictionResponse(
            campaign_id=prediction_request.campaign_id,
            predicted_conversions=prediction["value"],
            predicted_conversion_rate=prediction["conversion_rate"],
            confidence=prediction["confidence"],
            factors=prediction["contributing_factors"],
        ),
    )


@router.get("/models/status")
async def get_model_status():
    """
    Get status of ML models.

    Returns information about loaded models and their versions.
    """
    from app.ml.inference import ModelRegistry

    registry = ModelRegistry()

    return APIResponse(
        success=True,
        data={
            "ml_provider": settings.ml_provider,
            "models": registry.get_model_info(),
            "is_healthy": registry.health_check(),
        },
    )

File: frontend/src/api/simulator.ts
/**
 * ADs Growth System - Simulator API
 *
 * What-If simulator, ROAS forecasting, and conversion predictions.
 */

import { useQuery, useMutation, useQueryClient } from '@tanstack/react-query'
import { apiClient, ApiResponse } from './client'

// Types
export interface SimulationRequest {
  campaign_id?: number
  budget_change_percent: number
  days_ahead?: number
  include_confidence_interval?: boolean
}

export interface SimulationResponse {
  campaign_id?: number
  current_metrics: Record<string, unknown>
  predicted_metrics: Record<string, unknown>
  budget_change_percent: number
  confidence_interval?: Record<string, unknown>
  feature_importances?: Record<string, unknown>
  model_version: string
}

export interface ROASForecastRequest {
  campaign_ids?: number[]
  days_ahead?: number
  granularity?: string
}

export interface ROASForecastResponse {
  forecasts: Record<string, unknown>[]
  model_version: string
  generated_at: string
}

export interface ConversionPredictionRequest {
  campaign_id: number
  features: Record<string, unknown>
}

export interface ConversionPredictionResponse {
  campaign_id: number
  predicted_conversions: number
  predicted_conversion_rate: number
  confidence: number
  factors: Record<string, unknown>[]
}

export interface ModelStatus {
  ml_provider: string
  models: Record<string, unknown>
  is_healthy: boolean
}

// API Functions
export const simulatorApi = {
  simulateScenario: async (data: SimulationRequest): Promise<SimulationResponse> => {
    const response = await apiClient.post<ApiResponse<SimulationResponse>>('/simulator', data)
    return response.data.data
  },

  forecastRoas: async (data: ROASForecastRequest): Promise<ROASForecastResponse> => {
    const response = await apiClient.post<ApiResponse<ROASForecastResponse>>('/simulator/forecast/roas', data)
    return response.data.data
  },

  predictConversions: async (data: ConversionPredictionRequest): Promise<ConversionPredictionResponse> => {
    const response = await apiClient.post<ApiResponse<ConversionPredictionResponse>>(
      '/simulator/predict/conversions',
      data
    )
    return response.data.data
  },

  getModelStatus: async (): Promise<ModelStatus> => {
    const response = await apiClient.get<ApiResponse<ModelStatus>>('/simulator/models/status')
    return response.data.data
  },
}

// React Query Hooks
export function useSimulateScenario() {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: simulatorApi.simulateScenario,
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['simulator'] })
    },
  })
}

export function useForecastRoas() {
  return useMutation({
    mutationFn: simulatorApi.forecastRoas,
  })
}

export function usePredictConversions() {
  return useMutation({
    mutationFn: simulatorApi.predictConversions,
  })
}

export function useModelStatus() {
  return useQuery({
    queryKey: ['simulator', 'models', 'status'],
    queryFn: simulatorApi.getModelStatus,
    staleTime: 60 * 1000,
  })
}
